src/lavinmq/vhost.cr

```diff
@@ -369,18 +369,54 @@ module LavinMQ
       upstreams.stop_all
     end
 
+    private def close_connections(reason)
+      WaitGroup.wait do |wg|
+        to_close = Channel(Client).new
+        fiber_count = 0
+        @connections.each do |client|
+          loop do
+            select
+            when to_close.send client
+              break
+            else
+              fiber_id = fiber_count &+= 1
+              @log.trace { "spawning close conn fiber #{fiber_id} " }
+              wg.spawn do
+                while client_to_close = to_close.receive?
+                  client_to_close.close(reason)
+                end
+                @log.trace { "exiting close conn fiber #{fiber_id} " }
+              end
+              Fiber.yield
+            end
+          end
+        end
+        to_close.close
+      end
+    end
+
     def close(reason = "Broker shutdown")
       @closed = true
       stop_shovels
       stop_upstream_links
-      Fiber.yield
-      @log.debug { "Closing connections" }
-      @connections.each &.close(reason)
-      # wait up to 10s for clients to gracefully close
-      100.times do
-        break if @connections.empty?
-        sleep 0.1.seconds
+
+      @log.info { "Closing connections" }
+      close_done = Channel(Nil).new
+
+      spawn do
+        close_connections reason
+        @log.debug { "Close sent to all connections" }
+        close_done.close
       end
+
+      select
+      when close_done.receive?
+        @log.info { "All connections closed gracefully" }
+      when timeout 5.seconds
+        @log.warn { "Timeout waiting for connections to close. #{@connections.size} left." }
+      end
```

Comment:
Ah, my bad, for some reason I was thinking this was in a loop, that we logged every 5s but didn't timeout until after like 30-60s. 

```suggestion
      6.times do |i|
        select
        when close_done.receive?
          @log.info { "All connections closed gracefully" }
          return
        when timeout 5.seconds
          @log.warn { "Waiting for #{@connections.size} connections to close gracefully, #{(5-i)*5} more seconds" }
        end
      end
```

Or we just resort to 10-15s timeout. 5s feels a bit too short when dealing with huge amount of connections.

Reply:
Yes, I think one longer timeout is fine.